### PR TITLE
feat: add cross domain support

### PR DESCRIPTION
**Pull request by closetool**
Signed-off-by: closetool <c299999999@qq.com>

Fix: #493

### CODE REVIEW

File: examples/rbac_wtih_cross_domain_policy.csv [hunk 1]
```diff
@@ -0,0 +1,8 @@
+p, super_admin, /pen/1, GET
+p, super_admin, /pen2/1, GET
+p, book_admin, book_group, GET
+p, pen_admin, pen_group, GET
+
+g, alice, super_admin, domain1
+g, bob, book_admin, domain2
+g, bob, alice, domain2, domain1
```
Comment (nodece) [inline examples/rbac_wtih_cross_domain_policy.csv:8]: invalid rule.

File: model/assertion.go [hunk 2]
```diff
@@ -76,9 +76,9 @@ func (ast *Assertion) buildRoleLinks(rm rbac.RoleManager) error {
 		if len(rule) < count {
 			return errors.New("grouping policy elements do not meet role definition")
 		}
-		if len(rule) > count {
-			rule = rule[:count]
-		}
+		//if len(rule) > count {
+		//	rule = rule[:count]
+		//}
```
Comment (nodece) [inline model/assertion.go:81]: rule length should equal count value, you should remove this comment.

File: rbac/default-role-manager/role_manager_test.go [hunk 3]
```diff
@@ -295,3 +295,16 @@ func TestMatchingFuncOrder(t *testing.T) {
...
 	testRole(t, rm, "u1", "g2", true)
 }
+
+func TestCrossDomainRole(t *testing.T) {
+	rm := NewRoleManager(10)
+
+	_ = rm.AddLink("u1", "g1", "domain1")
+	_ = rm.AddLink("g1","g3","domain1")
+	_ = rm.AddLink("u2", "g2", "domain2")
+	_ = rm.AddLink("u1","u2","domain1", "domain2")
```
Comment (hsluoyz) [inline rbac/default-role-manager/role_manager_test.go:305]: Use gofmt

File: errors/rbac_errors.go [hunk 4]
```diff
@@ -20,6 +20,7 @@ import "errors"
 var (
 	ERR_NAME_NOT_FOUND            = errors.New("error: name does not exist")
 	ERR_DOMAIN_PARAMETER          = errors.New("error: domain should be 1 parameter")
+	ERR_MAX_DOMAIN_PARAMETER = errors.New("error: domain should most be 2 parameter")
```
Comment (hsluoyz) [inline errors/rbac_errors.go:23]: gofmt

File: rbac/default-role-manager/role_manager.go [hunk 5]
```diff
@@ -84,11 +84,64 @@ func (rm *RoleManager) Clear() error {
 func (rm *RoleManager) AddLink(name1 string, name2 string, domain ...string) error {
 	if len(domain) == 0 {
 		domain = []string{defaultDomain}
-	} else if len(domain) > 1 {
-		return errors.ERR_DOMAIN_PARAMETER
```
Comment (hsluoyz) [inline rbac/default-role-manager/role_manager.go:88]: Is `ERR_DOMAIN_PARAMETER` still in use?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by hsluoyz** [comment]
@rrasulzade plz review.

File: rbac/default-role-manager/role_manager.go (hunk 5 above)
Comment (closetool) [reply rbac/default-role-manager/role_manager.go:88]: yes, it it.

File: model/assertion.go (hunk 2 above)
Comment (hsluoyz) [inline model/assertion.go:81]: why?

File: examples/rbac_wtih_cross_domain_policy.csv (hunk 1 above)
Comment (hsluoyz) [inline examples/rbac_wtih_cross_domain_policy.csv:8]: alice is not a role name

File: rbac/default-role-manager/role_manager.go [hunk 6]
```diff
@@ -176,9 +195,9 @@ func (rm *RoleManager) HasLink(name1 string, name2 string, domain ...string) (bo
 
 		roles, _ := rm.allDomains.LoadOrStore(domain, new(Roles))
 		allRoles = roles.(*Roles)
-		if !allRoles.hasRole(name1, rm.matchingFunc) || !allRoles.hasRole(name2, rm.matchingFunc) {
-			continue
-		}
+		//if !allRoles.hasRole(name1, rm.matchingFunc) || !allRoles.hasRole(name2, rm.matchingFunc) {
+		//	continue
+		//}
```
Comment (hsluoyz) [inline rbac/default-role-manager/role_manager.go:200]: why?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: model/assertion.go (hunk 2 above)
Comment (closetool) [reply model/assertion.go:81]: cause it blocks the second domain

File: rbac/default-role-manager/role_manager.go (hunk 6 above)
Comment (closetool) [reply rbac/default-role-manager/role_manager.go:200]: if we support cross domain role inheriting, the two roles are not in the same domain

**Comment by closetool** [comment]
@hsluoyz Is this looks better?

File: rbac/default-role-manager/role_manager_test.go [hunk 7]
```diff
@@ -295,3 +295,18 @@ func TestMatchingFuncOrder(t *testing.T) {
...
 	testRole(t, rm, "u1", "g2", true)
 }
+
+func TestCrossDomainRole(t *testing.T) {
+	rm := NewRoleManager(10)
+
+	_ = rm.AddLink("super_admin", "book_admin", "domain1")
+	_ = rm.AddLink("super_admin", "pen_admin", "domain2")
+	_ = rm.AddLink("alice", "book_admin", "domain2")
+	_ = rm.AddLink("bob", "super_admin", "domain1", "domain2")
```
Comment (rrasulzade) [inline rbac/default-role-manager/role_manager_test.go:305]: does this line mean that bob is a super admin in domain1 and domain2? based on the tests below I see the opposite

Comment (closetool) [reply rbac/default-role-manager/role_manager_test.go:305]: No, it means bob in domain1 inherits super_admin in domain2, but bob has not role of super_admin in domain1.

File: errors/rbac_errors.go [hunk 8]
```diff
@@ -20,6 +20,7 @@ import "errors"
 var (
 	ERR_NAME_NOT_FOUND            = errors.New("error: name does not exist")
 	ERR_DOMAIN_PARAMETER          = errors.New("error: domain should be 1 parameter")
+	ERR_MAX_DOMAIN_PARAMETER      = errors.New("error: domain should most be 2 parameter")
```
Comment (hsluoyz) [inline errors/rbac_errors.go:23]: error: domain should be 2 parameters at most

File: rbac/default-role-manager/role_manager_test.go (hunk 7 above)
Comment (hsluoyz) [reply rbac/default-role-manager/role_manager_test.go:305]: @closetool here the role logic is very complicated. I think you can draw a role graph like what I did before and put it as a comment, so everyone can understand it easily:

https://github.com/casbin/casbin/blob/0c7aac93d766aeddea324d7a16fd8be1c700bca5/rbac/default-role-manager/role_manager_test.go#L171-L176

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by nodece** [comment]
could you describe the cross domain design in model? `g, bob, super_admin, domain1, domain2` seems that bob is a super admin in domain1 and domain2, as @rrasulzade said.

**Comment by hsluoyz** [comment]
This PR is not mature, closed for now.

**Pull request closed without merging**